harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -28,56 +31,199 @@
     pass
 
 
-class PyTorchTrialController(det.TrialController):
-    def __init__(self, trial_inst: det.Trial, *args: Any, **kwargs: Any) -> None:
-        super().__init__(*args, **kwargs)
+class TrainUnit:
+    def __init__(self, value: int):
+        self.value = value
+
+    @staticmethod
+    def _from_searcher_unit(length: int, unit: core.Unit):
+        if unit == core.Unit.EPOCHS:
+            return Epoch(length)
+        elif unit == core.Unit.RECORDS:
+            return Record(length)
+        elif unit == core.Unit.BATCHES:
+            return Batch(length)
+        else:
+            raise ValueError(f"unrecognized searcher unit {unit}")
+
+    @staticmethod
+    def _from_values(
+        batches: Optional[int] = None, records: Optional[int] = None, epochs: Optional[int] = None
+    ):
+        if sum((batches is not None, records is not None, epochs is not None)) != 1:
+            raise ValueError(f"invalid length: batches={batches} records={records} epochs={epochs}")
+        if batches:
+            return Batch(batches)
+        elif records:
+            return Record(records)
+        elif epochs:
+            return Epoch(epochs)
+
+    def _divides(self, steps: int):
+        assert self.value > 0, "TrainUnit value must be > 0"
+        if steps == 0:
+            return False
+        return steps % self.value == 0
+
+
+class Epoch(TrainUnit):
+    pass
+
+
+class Batch(TrainUnit):
+    pass
+
+
+class Record(TrainUnit):
+    pass
+
+
+class _TrainStepType(enum.Enum):
+    CHECKPOINT = "CHECKPOINT"
+    REPORT = "REPORT"
+    VALIDATE = "VALIDATE"
+    TRAIN = "TRAIN"
+
+
+# XXX: should this be private? will be returned from .train() but is only used internally
+class _TrainStep:
+    def __init__(self, step_type: _TrainStepType, unit: TrainUnit):
+        self.step_type = step_type
+        self.unit = unit
+        self.limit_reached = False
+
+
+class ShouldExit(Exception):
+    """
+    ShouldExit breaks out of the top-level train loop from inside function calls.
+    """
+
+    def __init__(self, skip_exit_checkpoint: bool = False):
+        self.skip_exit_checkpoint = skip_exit_checkpoint
+
+
+class _TrialState:
+    def __init__(
+        self,
+        trial_id: int = 0,
+        last_ckpt: int = 0,
+        steps_completed: int = 0,
```

Comment:
if steps_completed is for legacy compatibility with the old wlsq checkpoints, then:
- explain that in a comment
- make sure you aren't emitting legacy values in newly-created checkpoints.

Reply:
done.

Reply:
done. steps_completed is now default None, and _load will check for legacy paths to convert steps_completed into new epoch/batch values.

Reply:
Why even have `steps_completed` in the `TrialState` at all?  Why not do the conversion right when you see a `wlsq.pkl` file, and let `TrialState` be up-to-date?

Reply:
remvoed